ratatool-scalacheck/src/main/scala/com/spotify/ratatool/scalacheck/AvroGenerator.scala

```diff
@@ -52,36 +53,41 @@ object AvroGeneratorOps extends AvroGeneratorOps
 trait AvroGeneratorOps {
   def specificRecordOf[A <: SpecificRecord: ClassTag]: Gen[A] = {
     val cls = implicitly[ClassTag[A]].runtimeClass.asInstanceOf[Class[A]]
-    val schema = SpecificData.get().getSchema(cls)
+    val specificCoder = AvroCoder.of(cls)
+    val genericCoder = AvroCoder.of(specificCoder.getSchema)
 
-    genericRecordOf(schema).map { generic =>
-      val writer = new GenericDatumWriter[GenericRecord](schema)
-      val out = new ByteArrayOutputStream()
-      val encoder = EncoderFactory.get().binaryEncoder(out, null)
-      writer.write(generic, encoder)
-      encoder.flush()
-
-      val bytes = out.toByteArray
-
-      val reader = new SpecificDatumReader(cls)
-      val decoder = DecoderFactory.get().binaryDecoder(bytes, null)
-
-      reader.read(null.asInstanceOf[A], decoder)
+    genericRecordOf(specificCoder.getSchema).map { generic =>
+      val bytes = CoderUtils.encodeToByteArray(genericCoder, generic)
+      CoderUtils.decodeFromByteArray(specificCoder, bytes)
     }
   }
 
   def genericRecordOf(schema: Schema): Gen[GenericRecord] = {
-    avroValueOf(schema).map { avroValue =>
+    /**
+     * Adding the Gen.const in front fixes behaviour with Gen.listOfN, which would
+     * previously return the exact result for every item in the list. Also means that the
+     * behaviour of the generated data matches that of other `Gen[T]`
+     */
+    Gen.const(0).flatMap( _ => avroValueOf(schema).map { avroValue =>
```

Comment:
@kanterov @nevillelyh If there's one part I would like feedback on it is this. Please take a look :)

Reply:
Can you clarify the `Gen.listOfN` behavior?

Reply:
`Gen.listOfN(100, genericRecordOf(mySchema))`
will return 100 copies of the same genericRecord

Reply:
The behaviour becomes dependent on how the data is generated and requires internal knowledge of how the seed is managed. As another example `Gen.sequence((1 to 100).map(_ => genericRecordOf(schema)))` will return 100 different records

Reply:
@idreeskhan The comment looks incomplete? `behavior of the generated..`

Reply:
Already fixed :) @ankscircle

Reply:
It feels like something is broken, that is not a surprise given how much mutability is there.